cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -595,7 +598,13 @@ struct GrouperFastImpl : public Grouper {
     return std::move(impl);
   }
 
-  ~GrouperFastImpl() { map_.cleanup(); }
+  Status Reset() override {
+    rows_.Clean();
+    rows_minibatch_.Clean();
+    map_.cleanup();
+    RETURN_NOT_OK(map_.init(encode_ctx_.hardware_flags, ctx_->memory_pool()));
```

Comment:
What about `dictionaries_`, `map_equal_impl_` and `map_append_impl_`? Do they need resetting as well?

Reply:
For `dictionaries_`, it seems that it is currently assumed to be identical across all the `ConsumeImpl` calls:
https://github.com/apache/arrow/blob/a715ea06b71ec206a987d7921264778e9954404b/cpp/src/arrow/compute/row/grouper.cc#L633-L637

And both `map_equal_impl_` and `map_append_impl_` are just lambdas capturing only `this` pointer.

Reply:
Can you perhaps add a comment about `dictionaries_` then?

Reply:
Sure, will do.

Reply:
Done.